### PR TITLE
C++: Support for & and >> in SimpleRangeAnalysis

### PR DESCRIPTION
**Pull request by geoffw0**
Add support for bitwise `&` and `>>` (by a constant) in SimpleRangeAnalysis.qll.  This will fix https://github.com/Semmle/ql/issues/3247.

I've done a bit of performance testing locally, and the query seems to be up to 2% slower in the cases where I can measure a difference.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 1]
```diff
@@ -118,36 +118,74 @@ private string getValue(Expr e) {
...
 }
 
+/**
+ * A bitwise `&` expression in which both operands are unsigned, or are effectively
+ * unsigned due to being a non-negative constant.
+ */
+private class UnsignedBitwiseAndExpr extends BitwiseAndExpr {
+  UnsignedBitwiseAndExpr() {
+    (
+      getLeftOperand().getType().getUnderlyingType().(IntegralType).isUnsigned() or
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:128]: Does this do the right thing with conversions? For example, does it handle `unsigned z = (unsigned)x & (unsigned)y`, where `x` and `y` are of type `int`?

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll]: Well spotted - it doesn't, it should be working off the fully converted expression as that's the true input to `&`.  I've pushed some tests and a fix.

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 2]
```diff
@@ -125,11 +125,16 @@ private string getValue(Expr e) {
 private class UnsignedBitwiseAndExpr extends BitwiseAndExpr {
   UnsignedBitwiseAndExpr() {
     (
-      getLeftOperand().getType().getUnderlyingType().(IntegralType).isUnsigned() or
+      getLeftOperand().getFullyConverted().getType().getUnderlyingType().(IntegralType).isUnsigned() or
       getLeftOperand().getValue().toInt() >= 0
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:129]:
```suggestion
      getLeftOperand().getFullyConverted().getValue().toInt() >= 0
```

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 3]
```diff
@@ -125,11 +125,16 @@ private string getValue(Expr e) {
...
     ) and
     (
-      getRightOperand().getType().getUnderlyingType().(IntegralType).isUnsigned() or
+      getRightOperand()
+          .getFullyConverted()
+          .getType()
+          .getUnderlyingType()
+          .(IntegralType)
+          .isUnsigned() or
       getRightOperand().getValue().toInt() >= 0
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:138]:
```suggestion
      getRightOperand().getFullyConverted().getValue().toInt() >= 0
```

**Comment by geoffw0** [comment]
Thanks @dbartol !

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 4]
```diff
@@ -794,6 +864,22 @@ private float getUpperBoundsImpl(Expr expr) {
...
+    left = getFullyConvertedUpperBounds(andExpr.getLeftOperand()) and
+    right = getFullyConvertedUpperBounds(andExpr.getRightOperand()) and
+    result = left.minimum(right)
+  )
+  or
+  // `>>` by a constant
+  exists(RShiftExpr rsExpr, float left, int right |
+    rsExpr = expr and
+    left = getFullyConvertedUpperBounds(rsExpr.getLeftOperand()) and
+    right = rsExpr.getRightOperand().getValue().toInt() and
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:880]: I think this should have `.getFullyConverted()`.

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 5]
```diff
@@ -794,6 +864,22 @@ private float getUpperBoundsImpl(Expr expr) {
...
+    result = left / 2.pow(right)
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:881]: I think this should be rounded down.

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll (hunk 4 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:880]: Fixed by https://github.com/github/codeql/pull/3925.

**Pull request merged**